File: be/app/schemas/article.py
from typing import List, Optional

from pydantic import BaseModel

from app.schemas.base import BaseSchema, TimestampMixin


class ArticleCategoryBase(BaseModel):
    name: str
    status: bool = True


class ArticleCategoryCreate(ArticleCategoryBase):
    pass


class ArticleCategoryUpdate(BaseModel):
    name: Optional[str] = None
    status: Optional[bool] = None


class ArticleCategory(ArticleCategoryBase, TimestampMixin, BaseSchema):
    id: int


class ArticleTagBase(BaseModel):
    name: str
    name_code: Optional[str] = None
    status: bool = True


class ArticleTagCreate(ArticleTagBase):
    pass


class ArticleTagUpdate(BaseModel):
    name: Optional[str] = None
    name_code: Optional[str] = None
    status: Optional[bool] = None


class ArticleTag(ArticleTagBase, TimestampMixin, BaseSchema):
    id: int


class ArticleBase(BaseModel):
    author_id: int
    title: str
    title_code: Optional[str] = None
    description: Optional[str] = None
    description_code: Optional[str] = None
    content: str
    country_id: Optional[int] = None
    region_id: Optional[int] = None
    district_id: Optional[int] = None
    ward_id: Optional[int] = None
    thumbnail_url: Optional[str] = None
    status: bool = True


class ArticleCreate(ArticleBase):
    categories: Optional[List[int]] = None
    tags: Optional[List[int]] = None


class ArticleUpdate(BaseModel):
    title: Optional[str] = None
    title_code: Optional[str] = None
    description: Optional[str] = None
    description_code: Optional[str] = None
    content: Optional[str] = None
    country_id: Optional[int] = None
    region_id: Optional[int] = None
    district_id: Optional[int] = None
    ward_id: Optional[int] = None
    thumbnail_url: Optional[str] = None
    view_count: Optional[int] = None
    status: Optional[bool] = None
    categories: Optional[List[int]] = None
    tags: Optional[List[int]] = None


class Article(ArticleBase, TimestampMixin, BaseSchema):
    id: int
    view_count: int = 0


class ArticleWithRelations(Article):
    categories: List[ArticleCategory] = []
    tags: List[ArticleTag] = []


class ArticleArticleCategoryBase(BaseModel):
    article_id: int
    article_categories_id: int


class ArticleArticleTagBase(BaseModel):
    article_id: int
    article_tags_id: int


class ArticleCommentBase(BaseModel):
    user_id: int
    article_id: int
    content: str
    status: bool = True


class ArticleCommentCreate(ArticleCommentBase):
    pass


class ArticleCommentUpdate(BaseModel):
    content: Optional[str] = None
    status: Optional[bool] = None


class ArticleComment(ArticleCommentBase, TimestampMixin, BaseSchema):
    id: int


class ArticleReactionBase(BaseModel):
    user_id: int
    article_id: int
    status: bool = True


class ArticleReactionCreate(ArticleReactionBase):
    pass


class ArticleReactionUpdate(BaseModel):
    status: Optional[bool] = None


class ArticleReaction(ArticleReactionBase, TimestampMixin, BaseSchema):
    id: int

File: java/src/main/java/com/williamtravel/app/entity/Article.java
package com.williamtravel.app.entity;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.LocalDateTime;
import java.util.HashSet;
import java.util.Set;

/**
 * Stores travel articles and blog posts
 */
@Entity
@Table(name = "article")
@Data
@NoArgsConstructor
@AllArgsConstructor
public class Article {

    /**
     * Unique identifier for each article
     */
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id")
    private Integer id;

    /**
     * Reference to the user who authored this article
     */
    @ManyToOne
    @JoinColumn(name = "author_id", nullable = false)
    private User author;

    /**
     * Article title
     */
    @Column(name = "title", length = 255, nullable = false)
    private String title;

    /**
     * URL-friendly slug for the article
     */
    @Column(name = "slug", length = 255, unique = true)
    private String slug;

    /**
     * Code for translation purposes
     */
    @Column(name = "title_code", length = 255)
    private String titleCode;

    /**
     * Short description or excerpt of the article
     */
    @Column(name = "description", columnDefinition = "TEXT")
    private String description;

    /**
     * Description code for translation
     */
    @Column(name = "description_code", columnDefinition = "TEXT")
    private String descriptionCode;

    /**
     * Full article content
     */
    @Column(name = "content", columnDefinition = "TEXT", nullable = false)
    private String content;

    /**
     * Reference to the country the article is about
     */
    @ManyToOne
    @JoinColumn(name = "country_id")
    private Country country;

    /**
     * Reference to the region the article is about
     */
    @ManyToOne
    @JoinColumn(name = "region_id")
    private Region region;

    /**
     * Reference to the district the article is about
     */
    @ManyToOne
    @JoinColumn(name = "district_id")
    private District district;

    /**
     * Reference to the ward the article is about
     */
    @ManyToOne
    @JoinColumn(name = "ward_id")
    private Ward ward;

    /**
     * URL to the thumbnail image
     */
    @Column(name = "thumbnail_url", length = 255)
    private String thumbnailUrl;

    /**
     * Number of times the article has been viewed
     */
    @Column(name = "view_count")
    private Integer viewCount;

    /**
     * Whether the article is active/published
     */
    @Column(name = "status", nullable = false)
    private Boolean status;

    /**
     * Timestamp when article was created
     */
    @Column(name = "created_at", nullable = false)
    private LocalDateTime createdAt;

    /**
     * Timestamp when article was last updated
     */
    @Column(name = "updated_at")
    private LocalDateTime updatedAt;
    
    /**
     * Categories associated with this article
     */
    @ManyToMany
    @JoinTable(
        name = "article_article_categories",
        joinColumns = @JoinColumn(name = "article_id"),
        inverseJoinColumns = @JoinColumn(name = "article_categories_id")
    )
    private Set<ArticleCategory> categories = new HashSet<>();
    
    /**
     * Tags associated with this article
     */
    @ManyToMany
    @JoinTable(
        name = "article_article_tags",
        joinColumns = @JoinColumn(name = "article_id"),
        inverseJoinColumns = @JoinColumn(name = "article_tags_id")
    )
    private Set<ArticleTag> tags = new HashSet<>();
    
    /**
     * Comments on this article
     */
    @OneToMany(mappedBy = "article")
    private Set<ArticleComment> comments = new HashSet<>();
    
    /**
     * Reactions (likes, dislikes) to this article
     */
    @OneToMany(mappedBy = "article")
    private Set<ArticleReaction> reactions = new HashSet<>();
}
